### ISSUE TITLE
Develop action_conditions

### ISSUE THREAD

**Issue by a-matai** [uchicago-cs/chiventure#315]
The goal of this issue is to take over the conditions code from #215 to break it into a smaller issue. The key structural difference is the condition and effect will hold the actual pointer to the next item.

These are the necessary changes that need to be made to the conditions/effects list:

condition_t {
item_t *item (for debugging purposes more than anything)
attr_t *attr_to_check (pointer to the actual attribute from the item)
attr_value expected_value (self explanatory)
}

A few key things to work on here: 
- check_condition will need to be modified
- write code for all_conditions_met and add_action_condition
- @liuyifan-eric should the condition pointer hold the next or will WDL load that in differently? I really don't get why we put pointers in the structs so will defer to you. (Also relevant to develop effects)

**Comment by Kate32815** [comment]
Here are the parameters WDL would like for *add_action_condition()* to facilitate the process of loading action conditions into game state structs:
*add_action_condtion()*
This function creates a new condition for an action (belonging to an item) and adds it to the list of conditions for that action.

Parameters:
- g: pointer to a game state game struct
- item_id: the ID of the item we are performing the action on
- action_name: the name of the action for which we are adding conditions
- cond_item_id: the ID of the item specified in the action condition
- attribute: the attribute of the item specified in action condition that needs to be validated
- value: the value of the above attribute that needs to be validated

Output:
- 0 upon SUCCESS
- 1 if the first item ID (item_id) does not exist
- 2 if the specified action (action_name) does not exist in the first item
- 3 if the second item ID (cond_item_id) does not exist

Side Effects:
- Creates an action condition from input parameters (cond_item_id, attribute, and value) and adds it to the condition list for the specified action in the specified item
`int add_action_condition(game_t* g, char* item_id, char* action_name, char* cond_item_id, char* attribute, char* value);`

**Comment by SnowballAntrobus** [comment]
For testing purposes, action management needs a function that adds to the condition list of a game_action.

**Comment by patrickkoopmans** [comment]
> For testing purposes, action management needs a function that adds to the condition list of a game_action.

Yep, we will be writing that as specified in the issue opening. Please check back tonight for what the parameters of the function will be!

**Comment by a-matai** [comment]
Hi @Kate32815, this is relevant to #316 as well.

A few questions. 

We were going to make an add_effect that looks like this. The add_condition is very similar.

/* add_action_effect creates an effect_t struct and adds it to the action pointed to
* Parameters:
* - ptr to action
* - ptr to item being modified
* - ptr to attribute to modify
* - new attribute_value_t (takes tag from attribute above)
* 
* Returns:
* int SUCCESS FAILURE
* error 2 if action NULL
* error 3 if item to modify is null
*/
int add_action_effect(game_action_t *action, item_t *item_to_modify, attribute_t *attribute, attribute_value_t *new_value);

Happy to make changes but a few things first.
Do you need it to be abstracted to include game? this is a big deal as we'd have to then move this function all the way up and don't see why it is necessary.
The way we've written it now is it will take in an action struct and an item_to_modify (for condition, this would be item_to_check). There is a function called get_item_from_room which should solve this.

We will also be changing it to include a pointer to the action struct. This can be easily done through get_action. 

Are these params as we have described them to you okay?

**Comment by a-matai** [comment]
@Kate32815 didn't clarify earlier but we can have it take in either ids or pointers. Just let us know if you really need it to take in a id/action char* and not structs.

@MTUC08

**Comment by jessxia** [comment]
Added header and implementation for add_action_condition:
/* add_action_condition() creates a new condition for an item's action and                        
 * adds to the action's condition list                                                            
 * Parameters:                                                                                    
 *  a pointer to the item the action is being performed on                                        
 *  a pointer to the action to which the condition is being added                                 
 *  a pointer to the item specified in the action condition                                       
 *  a pointer to the attribute of the item specified in action condition that needs to be validated                                                                                                
 *  a pointer to the value of the above attribute that needs to be validated                      
 * Returns:                                                                                       
 *  SUCCESS upon SUCCESS, 1 if first item ID does not exist                                       
 *  2 if specified action does not exist in first item                                            
 *  3 if second item ID does not exist                                                            
 */
int add_action_condition(item_t *item, game_action_t *action, item_t *cond_item, attribute_t *cond_attribute, attribute_value_t *cond_value);

I have not yet tested the function, but it compiles. Let me know if the parameters need to be changed!

**Comment by jessxia** [comment]
Added condition_new, which creates a the game_action_condition struct and is called within add_action_condition. Changed add_action_condition to use LL_APPEND instead of hardcoding linked list iterations. Completed all_conditions_met. Also removed the hidden double pointer within the effects and conditions linked list type definitions.

**Comment by Kate32815** [comment]
@jessxia the function looks good mostly! One small thing: would the function be able to take in an attribute name (a char*) instead of an attribute_t pointer?
Or is there a function that would help me get an attribute_t* given an item_t* and an attribute name?

**Comment by Kate32815** [comment]
Hi! There's a problem with us trying to find the condition item_t pointer, as we do not know which room it is in given our WDL file, so we cannot call get_item_in_room on it.

Is it possible to have a get_item_from_game function? (if it's not too much trouble, as it's quite cumbersome for us to look through all the items we have for which room the cond_item is in)

**Comment by patrickkoopmans** [comment]
> Hi! There's a problem with us trying to find the condition item_t pointer, as we do not know which room it is in given our WDL file, so we cannot call get_item_in_room on it.
> 
> Is it possible to have a get_item_from_game function? (if it's not too much trouble, as it's quite cumbersome for us to look through all the items we have for which room the cond_item is in)

Hi Kate, before we attempt to implement this, what do you suggest the parameters of get_item_from-game should be?

**Comment by Kate32815** [comment]
The parameters would be the pointer to a game struct and an item id in the form of a string
`get_item_from_game(game_t *g, char* item_id)`

**Comment by patrickkoopmans** [comment]
> The parameters would be the pointer to a game struct and an item id in the form of a string
> `get_item_from_game(game_t *g, char* item_id)`

That sounds doable. Just so you know, if we choose to implement this then you would also be responsible for loading every single item into a new hash table of items within the game struct (we can provide functions to help with this) and we would also not be able to have two items with the same name. This means that, for example, every single door in the game will have to have a different name. Let me know if this is okay!

**Comment by Kate32815** [comment]
in our WDL spec we already say that no two items can have the same name so that condition would be okay! I need to check in with Shriya who's working on load_items to see if she's okay with loading the new hash table of items

**Comment by patrickkoopmans** [comment]
> in our WDL spec we already say that no two items can have the same name so that condition would be okay! I need to check in with Shriya who's working on load_items to see if she's okay with loading the new hash table of items

Hi Kate, my bad. I just figured out a way that we can do this without needing to change load_items. Sorry for the premature reply--no need to reach out to Shriya. I have deleted my comment to avoid confusion later on.

**Comment by Kate32815** [comment]
Shriya's okay with adding a bit to load_items so we're waiting on a new function that adds an item to a new hash table of items within the game struct

 Parameters:
- g : pointer to a game struct
- item: pointer to an item struct

Returns:
- 0 on SUCCESS
- 1 otherwise
`add_item_to_game(game_t* g, item_t* item)`

**Comment by patrickkoopmans** [comment]
> Shriya's okay with adding a bit to load_items! So we're waiting on a new function **add_item_to_game()** that adds an item to a new hash table of items within the game struct:
> 
> Parameters:
> 
> * g : pointer to a game struct
> * item: pointer to an item struct
> 
> Returns:
> 
> * 0 on SUCCESS
> * 1 otherwise
> 
> `int add_item_to_game(game_t* g, item_t* item)`
> 
> (we anticipate it to be similar to add_room_to_game)
> 
> and then there would be a corresponding **get_item_from_game()** function as follows:
> 
> Parameters:
> 
> * g : pointer to a game struct
> * item_id : ID of item to be found in the form of a string
> 
> Returns:
> 
> * pointer to item_t struct if item is found
> * NULL if item is not found
> 
> `item_t* get_item_from_game(game_t *g, char* item_id)`
> 
> @patrickkoopmans please let me know if this is okay!

Sounds great! I will implement this soon, but you can expect the interface to be exactly as you describe.

**Comment by jessxia** [comment]
Added an initial version of an add_condition function similar to add_effect into game module. This will need to be changed to take in different parameters once the item hash table is implemented (since room will no longer be necessary).

**Comment by jessxia** [comment]
Edited add_condition function to no longer take in room_id. Function now calls get_item_from_game to access item pointers from the game struct's new item hash table. @Kate32815 Does this function work with your code now?

**Comment by Kate32815** [comment]
Looks good! Yes, this greatly facilitates our loading process. One small thing that won't really affect how we use it-- can the return int values for add_condition be changed to be from 0-4? (unless there is something else that should be returning 1, as skipping numbers in the current implementation doesn't make much sense otherwise)

**Comment by jessxia** [comment]
@Kate32815 Thanks for catching that: I'm not sure why I forgot 1 existed. 
add_condition and add_effect now return 0 through 4.

**Comment by jessxia** [comment]
Closing statement:
We continued our work on the game-state/develop_actions branch and finalized the game_action_condition struct to be:

- a pointer to the item whose attribute we will be either looking at
- a pointer to the attribute of the item that we want to check
- the expected value of the attribute
- a pointer to the next condition in the list (it was confirmed that we needed this in order to use utlist macros)

We added or edited the following functions, mostly in game_action.c, to support use of our condition struct:

- check_condition, which checks if a condition is met
- all_conditions_met, which checks if all conditions listed in an action are met
- condition_new, which mallocs space for a new condition and initializes it with given inputs
- add_action_condition, which adds a condition to an action's condition linked list and calls condition_new
- add_condition, a more abstracted version of the above function that takes in a game struct as well (in game.c)

If any further functions become necessary, we will implement them in a new issue.

**Comment by a-matai** [comment]
Ah they didn't return 1 because 1 was reserved for failure
________________________________

**Comment by jessxia** [comment]
Fixed. Documentation for add_condition/add_effect now say return FAILURE if add_action_condition/effect fail and returns 2-5 for other errors.

**Comment by lydiafilipe** [comment]
Issue Score: ✓+

Comments:
Nice work! Great discussions
